extendr-api/src/robj/mod.rs

```diff
@@ -734,65 +733,143 @@ where
 {
     fn as_typed_slice(&self) -> Option<&'a [T]>
     where
-        Self: 'a,
-    {
-        None
-    }
+        Self: 'a;
 
     fn as_typed_slice_mut(&mut self) -> Option<&'a mut [T]>
     where
-        Self: 'a,
-    {
-        None
+        Self: 'a;
+}
+
+/// Trait for providing a rust typed pointer to an r type.
+/// R data are hidden behind [`SEXP`] opaque pointers with an identifier
+/// for the underlying r storage type. Thus this trait is only meaningful for
+/// [`SEXP`]-wrappers.
+trait SexpAs {
+    /// Identifier for the R object modes / types represented as a value.
+    /// See [`Rtype`] for reference.
+    const R_TYPE_ID: u32;
+    /// Corresponding rust type to `XP`.
+    type Type;
+    /// Returns a typed pointer to the underlying value
+    fn as_(x: SEXP) -> *mut Self::Type;
+}
+
+impl SexpAs for Rbool {
+    /// Corresponds to `logical` in R
```

Comment:
I think these comments should either be comments (`//`) or impl-level rustdocs (above the `impl`)

Reply:
Why? The'll be annotated on the assoc. type in the docs properly. 
I truly dislike `//` usage for top-level items...

Reply:
Oh, I thought you intended this comment to be at struct level. If it's supposed to be just for the `R_TYPE_ID` then this is fine.